python/arcticdb/version_store/_store.py

```diff
@@ -2278,6 +2278,80 @@ def scan_object_sizes(self):
                     key, value[0], format_bytes(value[1]), format_bytes(value[1] / value[0])
                 )
             )
+            
+    def is_symbol_fragmented(self, symbol: str, segment_size: int = None) -> bool:
+        """
+        Check whether the number of segments that would be reduced by compaction is more than or equal to the
+        value specified by the configuration option "SymbolDataCompact.SegmentCount" (defaults to 100).
+
+        Parameters
+        ----------
+        symbol: `str`
+            Symbol name.
+        segment_size: `int`
+            Target for maximum no. of rows per segment, after compaction.
+            If parameter is not provided, library option for segments's maximum row size will be used
+
+        Notes
+        ----------
+        Config map setting - SymbolDataCompact.SegmentCount will be replaced by a library setting
+        in the future. This API will allow overriding the setting as well.
+        
+        Returns
+        -------
+        bool
+        """
+        return self.version_store.is_symbol_fragmented(symbol, segment_size)
+
+    def defragment_symbol_data(self, symbol: str, segment_size: int = None) -> VersionedItem:
+        """
+        Compacts fragmented segments by merging row-sliced segments (https://docs.arcticdb.io/technical/on_disk_storage/#data-layer).
+        This method calls `is_symbol_fragmented` to determine whether to proceed with the defragmentation operation. 
+
+        CAUTION - One caveat of this method is that the resulting data will not be column sliced which may impact the performance
```

Comment:
Wait, if we then `update/append` data to a compacted symbol, aren't we creating that column groups size mismatch situation we say we aren't going to support?

Reply:
What does merged mean in this scenario?

Reply:
Remove column slicing.
Steps for defragmentation:
1. Remove column slicing
2. Remove row slicing as much as possible (per library setting `segment_row_size`)

Reply:
Remove column slicing.
Steps for defragmentation:
1. Remove column slicing
2. Remove row slicing as much as possible (per library setting `segment_row_size`)

Reply:
Ok - I think this will hit the case that Gem hit. The end result is that QueryBuilder queries breaks.

So just to clarify - we defragment the data, so we'll write out a new index key that no column slicing. If we append data, those rows will be column sliced. As a result the column slicing will differ from one row to another in the same index key, which I believe is the issue that @GemDot saw.

Reply:
At the very least we should add a test for this. Defragement, append, query using QueryBuilder.